source/grid/grid_reordering.cc

```diff
@@ -29,585 +29,710 @@
+            adjacent_cells[i] = invalid_cell;
+          }
+      }
 
-    MSide::MSide (const unsigned int initv0,
-                  const unsigned int initv1)
-      :
-      v0(initv0), v1(initv1),
-      Q0(numbers::invalid_unsigned_int),
-      Q1(numbers::invalid_unsigned_int),
-      lsn0(numbers::invalid_unsigned_int),
-      lsn1(numbers::invalid_unsigned_int),
-      Oriented(false)
-    {}
+      /**
+       * Comparison operator for edges. It compares based on the
+       * lexicographic ordering of the two vertex indices.
+       */
+      bool operator< (const Edge<dim> &e) const
+      {
+        return ((vertex_indices[0] < e.vertex_indices[0])
+                ||
+                ((vertex_indices[0] == e.vertex_indices[0]) && (vertex_indices[1] < e.vertex_indices[1])));
+      }
 
+      /**
+       * Compare two edges for equality based on their vertex indices.
+       */
+      bool operator== (const Edge<dim> &e) const
+      {
+        return ((vertex_indices[0] == e.vertex_indices[0])
+                &&
+                (vertex_indices[1] == e.vertex_indices[1]));
+      }
 
+      /**
+       * Store the given cell index as one of the cells as adjacent to
+       * the current edge. Also store that the current edge has index
+       * @p edge_within_cell within the given cell.
+       */
+      void add_adjacent_cell (const unsigned int cell_index,
+                              const unsigned int edge_within_cell)
+      {
+        const AdjacentCell adjacent_cell = { cell_index, edge_within_cell };
+        if (adjacent_cells[0].cell_index == numbers::invalid_unsigned_int)
+          adjacent_cells[0] = adjacent_cell;
+        else
+          {
+            Assert (adjacent_cells[1].cell_index == numbers::invalid_unsigned_int,
+                    ExcInternalError());
+            adjacent_cells[1] = adjacent_cell;
+          }
+      }
 
-    bool
-    MSide::operator == (const MSide &s2) const
-    {
-      if ((v0 == s2.v0)&&(v1 == s2.v1))
-        {
-          return true;
-        }
-      if ((v0 == s2.v1)&&(v1 == s2.v0))
-        {
-          return true;
-        }
-      return false;
-    }
+      /**
+       * The global indices of the two vertices that bound this edge. These
+       * will be ordered so that the first index is less than the second.
+       */
+      unsigned int vertex_indices[2];
 
+      /**
+       * An enum that indicates the direction of this edge with
+       * regard to the two vertices that bound it.
+       */
+      enum OrientationStatus
+      {
+        not_oriented,
+        forward,
+        backward
+      };
 
-    bool
-    MSide::operator != (const MSide &s2) const
-    {
-      return !(*this == s2);
-    }
+      OrientationStatus orientation_status;
 
+      /**
+       * A structure that stores how this edge relates to the at most
+       * two cells that are next to it.
+       */
+      struct AdjacentCell
+      {
+        unsigned int cell_index;
```

Comment:
You could add a constructor that makes this struct invalid and rip out the initialization you do above.